README.md
# nanoPhotosProvider
### :white_circle: extension for nanoGALLERY
    
  
Publish your self-hosted photos simply and automatically to nanoGALLERY.  
Content is provided on demand when browsing the albums of the gallery.  
Generates thumbnails automatically.  

To be used as an add-on for nanoGALLERY (http://nanogallery.brisbois.fr).

### :white_circle: Usage

##### :one: Step 1: installation

On your webserver:
- create a folder named `nanoPhotosProvider` where you want to store your photos
- in this folder:
  - copy the files:
    - `nanoPhotosProvider.php`,
    - `nanoPhotosProvider.class.php`,
    - `nanoPhotosProvider.cfg` and
    - `nanoPhotosProvider.Encoding.php`
  - create a folder named `nanoPhotosContent`  
    - copy your photos here  
    - you can organize your photos in folders (= albums)  
    - in the filenames, you can separate title and description with `$$`  
    - add leading `@@@@@` to the filename of the images to be used as album covers  
  - edit the `nanoPhotosProvider.cfg` file for custom settings  

<br />
##### :two: Step 2: configure your HTML page

- The page can be located anywhere on your webserver.
- Install nanoGALLERY for jQuery (see http://nanogallery.brisbois.fr)
- Configure the call to the plugin:
  - Use the specific parameters: `kind` and `jsonProvider`
    - `kind`: set value to `'json'`
    - `jsonProvider`: URL to the `nanoPhotosProvider.php` file installed in step 1

Example:

```js
    jQuery(document).ready(function () {
      jQuery("#nanoGallery1").nanoGallery({
        thumbnailWidth: 150,
        thumbnailHeight: 150,
        kind: 'json',
        jsonProvider: 'http://mywebsever.com/mypath/nanoPhotosProvider/nanoPhotosProvider.php',
        locationHash: false
      });
    });
```
<br />
<br />
##### :three: Step 3: test your page to see the result ;-)

<br />
##### :four: Step 4: add/change content
Add files and folders, or renaname them.
Please note that the generated thumbnails are never purged, so you may delete the `_thumbnails` folders to force a new generation.
  
  
  
### :white_circle: Custom configuration
Custom settings are defined in `nanoPhotosProvider.cfg`

Section | Option | default value | Description
------------ | ------------- | ------------ | -------------
config | fileExtensions | "jpg\|jpeg\|png\|gif" | Supported file extensions
 | contentFolder | "/nanoPhotosContent" | Folder where albums and images are stored
 | sortOrder | "asc" | Filename sort ordr
 | titleDescSeparator | "$$" | Separator between title and description in the name of the file
 | albumBlackListDetector | "!!!!!" | Ignore albums (folders) containing this sequence in their name
 | imageBlackListDetector | "_hidden" | Ignore images containing this sequence in their name
 | albumCoverDetector | "@@@@@" | Leading sequence in the filename of the image to be used as an album cover  
thumbnailSizes | width | 150 | Thumbnail width in pixels (note: the value must match the value of `thumbnailWidth` in your nanoGALLERY)
 | height | 150 | Thumbnail height in pixels (note: the value must match the value of `thumbnailHeight` in your nanoGALLERY)
 | crop | false | Crop images to avoid black borders
 

<br />
### :white_circle: Supported image formats
JPEG, GIF and PNG.

<br />

### :warning: limitations
- The nanoGALLERY option `locationHash` should not be enabled if albums have more than 2 levels.  
- multiple thumbnail resolutions are not supported.  
  
### :copyright: License
nanoPhotosProvider is licensed under [CC BY-NC 3.0](http://creativecommons.org/licenses/by-nc/3.0/).  
Only for personal, non-profit organizations, or open source projects (without any kind of fee), you may use nanoPhotosProvider for free.


### :white_circle: Requirements
* nanoGALLERY (http://nanogallery.brisbois.fr)
* Webserver
* PHP >= v5.2 with GD-Library


changelog.md
nanoPhotosProvider - nanoGALLERY plugin
===========

**Visit nanoPhotosProvider homepage for usage details: [https://github.com/Kris-B/nanoPhotosProvider](https://github.com/Kris-B/nanoPhotosProvider)**


ChangeLog 
------

v0.9.3
------

- nanoGALLERY support for options `albumList`, `whiteList`, `blackList`, `album`  
- folders without images are ignored  
- nanoGALLERY v5.10.0 included in the demonstration folder    


v0.9.2
------

- albums are now displayed before images  
- new option `imageBlackListDetector`: images containing this string are ignored  
- nanoGALLERY v5.9.1 included in the demonstration folder  
Thanks to EelcoA (https://github.com/EelcoA) and to eae710 (https://github.com/eae710) for their contribution.

v0.9.1
------

- Fixed incorret nanoGALLERY version (now v5.9.0beta1)

v0.9.0
------
- fixed incompatibility with latest nanoGALLERY versions  
- please use nanoGALLERY v5.8.0beta1 (located in the demonstration folder)  
Thanks to Ruplahlava (https://github.com/Ruplahlava) for his contribution.

galleryJSON.php
<?php
/**
* galleryJSON add-on for nanoGALLERY
*
* Publish a folder/content structure in JSON for albums and images to be displayed with the jQuery plugin nanoGALLERY.
* This is an add-on for nanoGALLERY.
* 
* License: For personal, non-profit organizations, or open source projects (without any kind of fee), you may use nanoGALLERY for free. 
* -------- ALL OTHER USES REQUIRE THE PURCHASE OF A PROFESSIONAL LICENSE.
*
*
* PHP 5.2+
* @version    0.1.0
* @author     Christophe BRISBOIS - http://www.brisbois.fr/
* @copyright  Copyright 2014
* @license    CC BY-NC 3.0
* @link       https://github.com/Kris-B/nanoGALLERY
*
*/


  // comment next line to enable warnings
  error_reporting(E_ERROR | E_PARSE);

  class nanogalleryData {
    var $fullDir='';
    //var $images;
    //var $URI;
  }

  class item {
    var $src='';
    var $srct='';
    var $title='';
    var $description='';
    var $ID='';
    var $albumID='0';
    var $kind='';      // album
  }

  class nanogalleryJSON {

    function __construct() {
    
      // Retrieve an album's thumbnail (folder)
      function GetAlbumThumbnail( $baseFolder ) {

        // look for thumbnail containing the album thumbnail detector string
        if( is_dir( $baseFolder.'/_thumbnails' ) ) {
          $files = glob($baseFolder.'/_thumbnails/'.$GLOBALS['albumThumbnailDetector'].'*.*');
          if( count($files) > 0 ) {
            return '_thumbnails/'.basename($files[0]);
          }
        }
        
        // fallback 1: use first found thumbnail in thumbnails folder
        if( is_dir( $baseFolder.'/_thumbnails' ) ) {
          $i=GetFirstFileFolder( $baseFolder.'/_thumbnails' );
          if( $i != '' ) {
          return '_thumbnails/'.$i;
          }
        }
        
        // fallback 2: use first found image in images folder
        if( is_dir( $baseFolder.'/_images' ) ) {
          return GetFirstFileFolder( $baseFolder.'_images' );
        }
      }

      
      // Retrieve the first image of one folder --> ALBUM THUMBNAIL
      function GetFirstFileFolder( $folder ) {
        $image='';

        $dh = opendir($folder);
        while (false !== ($filename = readdir($dh))) {
          if ( is_file( $folder.'/'.$filename ) && preg_match("/\.(".$GLOBALS['fileExtensions'].")*$/i", $filename) ) {
            $image=$filename;
            break;
          }
        }
        closedir( $dh );
        
        return $image;
      }
      
      
      // Extract title and description from filename
      function GetTitleDesc( $filename, $isImage ) {
        $oneItem = new item();
        if( strpos($filename,$GLOBALS['titleDescSeparator']) > 0 ) {
          $s=explode($GLOBALS['titleDescSeparator'],$filename);
          $oneItem->title=utf8_encode($s[0]);
          if( $isImage ) {
            $oneItem->description=utf8_encode(preg_replace('/.[^.]*$/', '', $s[1]));
          }
          else {
            $oneItem->description=utf8_encode($s[1]);
          }
        }
        else {
          if( $isImage ) {
            $oneItem->title=utf8_encode(preg_replace('/.[^.]*$/', '', $filename));
          }
          else {
            $oneItem->title=utf8_encode($filename);
          }
        }
        return $oneItem;

      }
      
      function CustomEncode( $s ) {
        // if ( !mb_check_encoding($s, 'UTF-8')) {
        if( $GLOBALS['isUTF8Encoded'] == true ) {
          return rawurlencode($s);
        }
        else {
          return utf8_encode($s);
        }
      }
      
      function CustomDecode( $s ) {
        // if ( !mb_check_encoding($s, 'UTF-8')) {
        if( $GLOBALS['isUTF8Encoded'] == true ) {
          return urldecode($s);
        }
        else {
          return utf8_decode($s);
        }
      }
      

      // retrieve the album ID in the URL
      $album='/';
      $albumID=$_GET["albumID"];
      if( !$albumID == '0' && $albumID != '' ) {
        // $album='/'.utf8_decode(urldecode($albumID)).'/';
        $album='/'.CustomDecode($albumID).'/';
      }

      $data = new nanogalleryData();
      $data->fullDir= __DIR__.'/ngcontent'.($album);

      $config = parse_ini_file( './galleryJSON.cfg' );
      $GLOBALS['fileExtensions'] = $config['images'];
      $GLOBALS['sortOrder'] = strtoupper($config['sortOrder']);
      $GLOBALS['titleDescSeparator'] = strtoupper($config['titleDescSeparator']);
      $GLOBALS['albumThumbnailDetector'] = strtoupper($config['albumThumbnailDetector']);
      $GLOBALS['isUTF8Encoded'] = $config['isUTF8Encoded'];
      
      $lstImages = array();
      $lstAlbums = array();

      
      $dh = opendir($data->fullDir);
      // loop folder
      if( $dh != false ) {
        while (false !== ($filename = readdir($dh))) {

          if ( $filename != '.' && $filename != '..' ) {

            if ( is_dir($data->fullDir.$filename) ) {

              if ( $filename == '_images' ) {
                
                // CONTENT OF THE FOLDER --> IMAGES and THUMBNAILS
                $di = $data->fullDir.'_images';
                $dih = opendir( $di );
                while (false !== ($image = readdir($dih))) {
                
                  if ( is_file( $di.'/'.$image ) && preg_match("/\.(". $GLOBALS['fileExtensions'] .")*$/i", $image) ) {
                    $oneItem = new item();
                    // $oneItem->src=rawurlencode(utf8_encode('ngcontent'.$album.'_images/'.$image));
                    // $oneItem->src=rawurlencode(('ngcontent'.$album.'_images/'.$image));
                    $oneItem->src=CustomEncode(('ngcontent'.$album.'_images/'.$image));
                    
                    $tn = $data->fullDir.'_thumbnails/'.$image;
                    if( file_exists($tn) ) {
                      // $oneItem->srct=rawurlencode(utf8_encode('ngcontent'.$album.'_thumbnails/'.$image));
                      // $oneItem->srct=(rawurlencode('ngcontent'.$album.'_thumbnails/'.$image));
                      $oneItem->srct=(CustomEncode('ngcontent'.$album.'_thumbnails/'.$image));
                    }
                    else {
                      $tn = $data->fullDir.'_thumbnails/'.$GLOBALS['albumThumbnailDetector'].$image;
                      if( file_exists($tn) ) {
                        // $oneItem->srct=rawurlencode(utf8_encode('ngcontent'.$album.'_thumbnails/'.$GLOBALS['albumThumbnailDetector'].$image));
                        // $oneItem->srct=(rawurlencode('ngcontent'.$album.'_thumbnails/'.$GLOBALS['albumThumbnailDetector'].$image));
                        $oneItem->srct=(CustomEncode('ngcontent'.$album.'_thumbnails/'.$GLOBALS['albumThumbnailDetector'].$image));
                      }
                      else {
                        // fallback: use image if thumbnail not found
                        $tn='ngcontent'.$album.'_images/'.$image;
                        $oneItem->srct=$oneItem->src;
                      }
                    }
                    $size = getimagesize($tn);
                    $oneItem->imgtWidth=$size[0];
                    $oneItem->imgtHeight=$size[1];

                    $oneItem->albumID=$albumID;
                    
                    $e=GetTitleDesc($image, true);
                    $oneItem->title=$e->title;
                    $oneItem->description=$e->description;
                    
                    $lstImages[] = $oneItem;
                  }
                }
                closedir( $dih );
              }
              else
                if ( $filename == '_thumbnails' ) {
                  // THUMBNAIL FOLDER --> ignore
                }
                else {
                  // ALBUMS
                  $oneItem = new item();
                  $oneItem->kind='album';

                  $e=GetTitleDesc($filename, false);
                  $oneItem->title=$e->title;
                  $oneItem->description=$e->description;
                  
                  $oneItem->albumID=$albumID;
                  if( $albumID == '0' || $albumID == '' ) {
                    // $oneItem->ID=rawurlencode(utf8_encode($filename));
                    $oneItem->ID=CustomEncode($filename);
                  }
                  else {
                    // $oneItem->ID=$albumID.rawurlencode(utf8_encode('/'.$filename));
                    $oneItem->ID=$albumID.CustomEncode('/'.$filename);
                  }
                  
                  $s= GetAlbumThumbnail( $data->fullDir.$filename );

                  $t='';
                  if( $albumID == '0' || $albumID == '' ) {
                    $t=$filename;
                  }
                  else {
                    $t=$album.'/'.$filename;
                  }
                  // $oneItem->srct=rawurlencode(utf8_encode('/ngcontent/'.$t.'/'.$s));
                  // $oneItem->srct=rawurlencode(('/ngcontent/'.$t.'/'.$s));
                  $oneItem->srct=CustomEncode(('/ngcontent/'.$t.'/'.$s));
                  
                  $lstAlbums[] = $oneItem;

                }
              
            }
          }
        }
        closedir( $dh );
      }

      function compare($a, $b) {
        $al = strtolower($a->title);
        $bl = strtolower($b->title);
        if ($al == $bl) {
            return 0;
        }
        $b=false;
        switch ( $GLOBALS['sortOrder'] ) {
          case 'DESC' :
            if( $al < $bl ) { $b = true; }
            break;
          case 'ASC':
          default:
            if( $al > $bl ) { $b = true; }
            break;
        }
        return ($b) ? +1 : -1;
      }
      
      // sort data
      usort($lstAlbums, "compare");
      usort($lstImages, "compare");
      
      header('Content-Type: application/json; charset=utf-8');

      // return the data
      $callback = $_GET['jsonp'];
      $output = json_encode( array_merge($lstAlbums, $lstImages) );
      if (isset($callback)) {
        echo $callback . '(' . $output . ')';     // return JSONP
      }
      else {
        echo $output;                             // return JSON
      }
    }

  }

  $t = new nanogalleryJSON();

  exit;
?>

nanoPhotosProvider.php
<?php

/**
 * nanoPhotosProvider add-on for nanoGALLERY
 *
 * This is an add-on for nanoGALLERY (image gallery - http://nanogallery.brisbois.fr).
 * This PHP application will publish your images and albums from a PHP webserver to nanoGALLERY.
 * The content is provided on demand, one album at one time.
 * Thumbnails are generated automatically.
 * 
 * License: For personal, non-profit organizations, or open source projects (without any kind of fee), you may use nanoGALLERY for free. 
 * -------- ALL OTHER USES REQUIRE THE PURCHASE OF A PROFESSIONAL LICENSE.
 *
 *
 * PHP 5.2+
 * @version       0.9.3
 * @author        Christophe BRISBOIS - http://www.brisbois.fr/
 * @Contributor   Ruplahlava - https://github.com/Ruplahlava
 * @Contributor   EelcoA  - https://github.com/EelcoA
 * @Contributor   eae710 - https://github.com/eae710
 * @copyright     Copyright 2015
 * @license       CC BY-NC 3.0
 * @link          https://github.com/Kris-B/nanoPhotosProvider
 * @Support       https://github.com/Kris-B/nanoPhotosProvider/issues
 *
 */
require './nanoPhotosProvider.JSON.class.php';

// Available values development, production
// Codeigniter env switch https://github.com/bcit-ci/CodeIgniter/

define('ENVIRONMENT', 'production');

switch (ENVIRONMENT) {
    case 'development':
        error_reporting(-1);
        ini_set('display_errors', 1);
        break;

    case 'production':
        ini_set('display_errors', 0);
        if (version_compare(PHP_VERSION, '5.3', '>=')) {
            error_reporting(E_ALL & ~E_NOTICE & ~E_DEPRECATED & ~E_STRICT & ~E_USER_NOTICE & ~E_USER_DEPRECATED);
        } else {
            error_reporting(E_ALL & ~E_NOTICE & ~E_STRICT & ~E_USER_NOTICE);
        }
        break;

    default:
        header('HTTP/1.1 503 Service Unavailable.', TRUE, 503);
        echo 'The application environment is not set correctly.';
        exit(1); // EXIT_ERROR
}

$t = new galleryJSON();

?>


nanoPhotosProvider.JSON.class.php
<?php

/**
 * nanoPhotosProvider add-on for nanoGALLERY
 *
 * This is an add-on for nanoGALLERY (image gallery - http://nanogallery.brisbois.fr).
 * This PHP application will publish your images and albums from a PHP webserver to nanoGALLERY.
 * The content is provided on demand, one album at one time.
 * Thumbnails are generated automatically.
 * 
 * License: For personal, non-profit organizations, or open source projects (without any kind of fee), you may use nanoGALLERY for free. 
 * -------- ALL OTHER USES REQUIRE THE PURCHASE OF A PROFESSIONAL LICENSE.
 *
 *
 * PHP 5.2+
 * @version       0.9.3
 * @author        Christophe BRISBOIS - http://www.brisbois.fr/
 * @Contributor   Ruplahlava - https://github.com/Ruplahlava
 * @Contributor   EelcoA  - https://github.com/EelcoA
 * @Contributor   eae710 - https://github.com/eae710
 * @copyright     Copyright 2015
 * @license       CC BY-NC 3.0
 * @link          https://github.com/Kris-B/nanoPhotosProvider
 * @Support       https://github.com/Kris-B/nanoPhotosProvider/issues
 *
 */
require './nanoPhotosProvider.Encoding.php';

class galleryData
{
    public $fullDir = '';

    //public $images;
    //public $URI;
}

class item
{
    public $src         = '';
    public $srct        = '';
    public $title       = '';
    public $description = '';
    public $ID          = '';
    public $albumID     = '0';
    public $kind        = '';      // album, image

}

class galleryJSON
{
    protected $config = array();
    protected $data;
    protected $albumID;
    protected $album;
            
    const CONFIG_FILE    = './nanoPhotosProvider.cfg';
    //const CONTENT_FOLDER = '/nanoPhotosContent';

    public function __construct()
    {
        // retrieve the album ID in the URL
        $this->album   = '/';
        $this->albumID = '';
        if (isset($_GET['albumID'])) {
            // $this->albumID = $_GET['albumID'];
            $this->albumID = rawurldecode($_GET['albumID']);
        }
        if (!$this->albumID == '0' && $this->albumID != '' && $this->albumID != null) {
            // $this->album='/'.utf8_decode(urldecode($this->albumID)).'/';
            $this->album = '/' . $this->CustomDecode($this->albumID) . '/';
        } else {
            $this->albumID = '0';
        }

        $this->data          = new galleryData();
        $this->setConfig(self::CONFIG_FILE);
        // $this->data->fullDir = __DIR__ . self::CONTENT_FOLDER . ($this->album);
        $this->data->fullDir = __DIR__ . ($this->config['contentFolder']) . ($this->album);

        $lstImages = array();
        $lstAlbums = array();
        
        $dh = opendir($this->data->fullDir);

        // loop the folder to retrieve images and albums
        if ($dh != false) {
          while (false !== ($filename = readdir($dh))) {
            $filenameU=strtoupper($filename);
            if (is_file($this->data->fullDir . $filename) ) {
              // it's a file
              if ($filename != '.' &&
                      $filename != '..' &&
                      $filename != '_thumbnails' &&
                      preg_match("/\.(" . $this->config['fileExtensions'] . ")*$/i", $filename) &&
                      strpos($filename, $this->config['imageBlackListDetector']) == false )
              {
                $lstImages[] = $this->prepare_data($filename, 'IMAGE');
              }
            }
            else {
              // it's a folder
              $files = glob($this->data->fullDir . $filename."/*.{".str_replace("|",",",$this->config['fileExtensions'])."}", GLOB_BRACE);    // to check if folder contains images
              if ($filename != '.' &&
                      $filename != '..' &&
                      $filename != '_thumbnails' &&
                      strpos($filename, $this->config['albumBlackListDetector']) == false && 
                      !empty($files) )
              {
                $lstAlbums[] = $this->prepare_data($filename, 'ALBUM');
              }
            }
          }
          closedir($dh);
        }

        // sort data
        usort($lstAlbums, array('galleryJSON','compare'));
        usort($lstImages, array('galleryJSON','compare'));

        // return the data
        header('Content-Type: application/json; charset=utf-8');
        $output = json_encode(array_merge($lstAlbums, $lstImages));     // UTF-8 encoding is mandatory
        if (isset($_GET['jsonp'])) {
            // return in JSONP
            echo $_GET['jsonp'] . '(' . $output . ')';
        } else {
            // return in JSON
            echo $output;
        }
    }

    
    
    protected function setConfig($filePath)
    {
        $config                                 = parse_ini_file($filePath, true);
        $this->config['contentFolder']          = $config['config']['contentFolder'];
        $this->config['fileExtensions']         = $config['config']['fileExtensions'];
        $this->config['sortOrder']              = strtoupper($config['config']['sortOrder']);
        $this->config['titleDescSeparator']     = $config['config']['titleDescSeparator'];
        $this->config['albumCoverDetector']     = $config['config']['albumCoverDetector'];
        $this->config['albumBlackListDetector'] = strtoupper($config['config']['albumBlackListDetector']);
        $this->config['imageBlackListDetector'] = $config['config']['imageBlackListDetector'];


        if (isset($config['thumbnailSizes']['width']) && isset($config['thumbnailSizes']['height'])) {
            $this->config['thumbnailsGenerate']       = true;
            $this->config['thumbnailSizes']['width']  = $config['thumbnailSizes']['width'];
            $this->config['thumbnailSizes']['height'] = $config['thumbnailSizes']['height'];
            if (isset($config['thumbnailSizes']['crop'])) {
                $this->config['thumbnailSizes']['crop'] = $config['thumbnailSizes']['crop'];
            } else {
                $this->config['thumbnailSizes']['crop'] = false;
            }
        } else {
            $this->config['thumbnailsGenerate'] = false;
        }
    }

    /**
     * RETRIEVE THE COVER IMAGE (THUMBNAIL) OF ONE ALBUM (FOLDER)
     * 
     * @param string $baseFolder
     * @return string
     */
    protected function GetAlbumCover($baseFolder)
    {

        // look for cover image
        $files = glob($baseFolder . '/' . $this->config['albumCoverDetector'] . '*.*');
        if (count($files) > 0) {
            $i = basename($files[0]);
            if (preg_match("/\.(" . $this->config['fileExtensions'] . ")*$/i", $i)) {
                $tn = $this->GetThumbnail($baseFolder, $i);
                if ($tn != '') {
                    return $tn;
                }
            }
        }

        // no cover image found --> use the first image for the cover
        $i = $this->GetFirstImageFolder($baseFolder);
        if ($i != '') {
            $tn = $this->GetThumbnail($baseFolder, $i);
            if ($tn != '') {
                return $tn;
            }
        }

        return '';
    }

    /**
     * Retrieve the first image of one folder --> ALBUM THUMBNAIL
     * 
     * @param string $folder
     * @return string
     */
    protected function GetFirstImageFolder($folder)
    {
        $image = '';

        $dh       = opendir($folder);
        while (false !== ($filename = readdir($dh))) {
            if (is_file($folder . '/' . $filename) && preg_match("/\.(" . $this->config['fileExtensions'] . ")*$/i", $filename)) {
                $image = $filename;
                break;
            }
        }
        closedir($dh);

        return $image;
    }

    /**
     * 
     * @param object $a
     * @param object $b
     * @return int
     */
    protected function compare($a, $b)
    {
        $al = strtolower($a->title);
        $bl = strtolower($b->title);
        if ($al == $bl) {
            return 0;
        }
        $b = false;
        switch ($this->config['sortOrder']) {
            case 'DESC' :
                if ($al < $bl) {
                    $b = true;
                }
                break;
            case 'ASC':
            default:
                if ($al > $bl) {
                    $b = true;
                }
                break;
        }
        return ($b) ? +1 : -1;
    }

    /**
     * RETRIEVE ONE IMAGE'S THUMBNAIL
     * 
     * @param type $baseFolder
     * @param type $filename
     * @return type
     */
    protected function GetThumbnail($baseFolder, $filename)
    {
        $tn = $baseFolder . '_thumbnails/' . $filename;

        if (file_exists($tn)) {
            //if( filemtime($tn) < filemtime($baseFolder.$filename) ) {
            // image file is older as the thumbnail file
            return '_thumbnails/' . $filename;
            //}
        }

        // generate the thumbnail
        $tn = $this->GenerateThumbnail($baseFolder, $filename);
        if ($tn != '') {
            return $tn;
        }

        // fallback: original image (no thumbnail)
        return $filename;
    }

    /**
     * GENERATE ONE THUMBNAIL
     * 
     * @param type $baseFolder
     * @param type $filename
     * @return string
     */
    protected function GenerateThumbnail($baseFolder, $filename)
    {
        if (!$this->config['thumbnailsGenerate'] == true ) {
            return '';
        }

        $td = $baseFolder . '/_thumbnails';
        if (!file_exists($td)) {
            mkdir($td, 0777, true);
        }
        //$orgImage = imagecreatefromjpeg($baseFolder.'/'.$filename);

        $size = getimagesize($baseFolder . '/' . $filename);
        switch ($size['mime']) {
            case 'image/jpeg':
                $orgImage = imagecreatefromjpeg($baseFolder . '/' . $filename);
                break;
            case 'image/gif':
                $orgImage = imagecreatefromgif($baseFolder . '/' . $filename);
                break;
            case 'image/png':
                $orgImage = imagecreatefrompng($baseFolder . '/' . $filename);
                break;
            default:
                return '';
                break;
        }
        $width  = $size[0];
        $height = $size[1];

        $tnFilename  = $baseFolder . '/_thumbnails/' . $filename;
        $thumbWidth  = $this->config['thumbnailSizes']['width'];
        $thumbHeight = $this->config['thumbnailSizes']['height'];

        $originalAspect = $width / $height;
        $thumbAspect    = $thumbWidth / $thumbHeight;

        if ($this->config['thumbnailSizes']['crop']) {
            // CROP THE IMAGE
            // some inspiration found in donkeyGallery (from Gix075) https://github.com/Gix075/donkeyGallery 
            if ($originalAspect >= $thumbAspect) {
                // If image is wider than thumbnail (in aspect ratio sense)
                $newHeight = $thumbHeight;
                $newWidth  = $width / ($height / $thumbHeight);
            } else {
                // If the thumbnail is wider than the image
                $newWidth  = $thumbWidth;
                $newHeight = $height / ($width / $thumbWidth);
            }

            $thumb = imagecreatetruecolor($thumbWidth, $thumbHeight);

            // Resize and crop
            imagecopyresampled($thumb, $orgImage, 0 - ($newWidth - $thumbWidth) / 2, // dest_x: Center the image horizontally
                    0 - ($newHeight - $thumbHeight) / 2, // dest-y: Center the image vertically
                    0, 0, // src_x, src_y
                    $newWidth, $newHeight, $width, $height);
        } else {
            // DO NOT CROP
            if ($originalAspect >= $thumbAspect) {
                $newHeight = $height / $width * $thumbWidth;
                $newWidth  = $thumbWidth;
            } else {
                $newWidth  = $width / $height * $thumbHeight;
                $newHeight = $thumbHeight;
            }

            $thumb = imagecreatetruecolor($newWidth, $newHeight);

            // Resize
            imagecopyresampled($thumb, $orgImage, 0, 0, 0, 0, $newWidth, $newHeight, $width, $height);
        }

        switch ($size['mime']) {
            case 'image/jpeg':
                imagejpeg($thumb, $tnFilename, 90);
                break;
            case 'image/gif':
                imagegif($thumb, $tnFilename);
                break;
            case 'image/png':
                imagepng($thumb, $tnFilename, 1);
                break;
        }

        return '_thumbnails/' . $filename;
    }

    /**
     * Extract title and description from filename
     * 
     * @param string $filename
     * @param boolean $isImage
     * @return \item
     */
    protected function GetTitleDesc($filename, $isImage)
    {
        if ($isImage) {
            $filename = $this->file_ext_strip($filename);
        }

        $oneItem = new item();
        if (strpos($filename, $this->config['titleDescSeparator']) > 0) {
          // title and description
          $s              = explode($this->config['titleDescSeparator'], $filename);
          $oneItem->title = $this->CustomEncode($s[0]);
          if ($isImage) {
            $oneItem->description = $this->CustomEncode(preg_replace('/.[^.]*$/', '', $s[1]));
          } else {
            $oneItem->description = $this->CustomEncode($s[1]);
          }
        } else {
          // only title
          if ($isImage) {
            $oneItem->title = $this->CustomEncode($filename);  //(preg_replace('/.[^.]*$/', '', $filename));
          } else {
            $oneItem->title = $this->CustomEncode($filename);
          }
          $oneItem->description = '';
        }

        $oneItem->title = str_replace($this->config['albumCoverDetector'], '', $oneItem->title);   // filter cover detector string
        return $oneItem;
    }

    /**
     * Returns only the file extension (without the period).
     * 
     * @param string $filename
     * @return string
     */
    protected function file_ext($filename)
    {
        if (!preg_match('/./', $filename)) {
            return '';
        }
        return preg_replace('/^.*./', '', $filename);
    }

    /**
     * Returns the file name, less the extension.
     * 
     * @param string $filename
     * @return string
     */
    protected function file_ext_strip($filename)
    {
        return preg_replace('/.[^.]*$/', '', $filename);
    }

    
    
    /**
     * 
     * @param string $s
     * @return string
     */
    protected function CustomEncode($s)
    {
        return \ForceUTF8\Encoding::toUTF8(($s));
        //return \ForceUTF8\Encoding::fixUTF8(($s));
    }

    /**
     * 
     * @param type $s
     * @return type
     */
    protected function CustomDecode($s)
    {
        return utf8_decode($s);
        // return $s;
    }

    protected function prepare_data($filename, $kind)
    {
        $oneItem = new item();
        // if (is_file($this->data->fullDir . $filename) && preg_match("/\.(" . $this->config['fileExtensions'] . ")*$/i", $filename)) {
        if ( $kind == 'IMAGE' ) {
            // ONE IMAGE

            $oneItem->kind = 'image';
            $e                    = $this->GetTitleDesc($filename, true);
            $oneItem->title       = $e->title;
            $oneItem->description = $e->description;
            $oneItem->src         = rawurlencode($this->CustomEncode($this->config['contentFolder'] . $this->album . '/' . $filename));

            $tn                  = $this->GetThumbnail($this->data->fullDir, $filename);
            $oneItem->srct       = rawurlencode($this->CustomEncode($this->config['contentFolder'] . $this->album . $tn));
            $size                = getimagesize($this->data->fullDir . $tn);
            $oneItem->imgtWidth  = $size[0];
            $oneItem->imgtHeight = $size[1];

            $oneItem->albumID = rawurlencode($this->albumID);
            if ($this->albumID == '0' || $this->albumID == '') {
                $oneItem->ID = rawurlencode($this->CustomEncode($e->title));
            } else {
                $oneItem->ID = rawurlencode($this->albumID . $this->CustomEncode('/' . $e->title));
            }
            return $oneItem;
        } else {
            // ONE ALBUM
            $oneItem->kind = 'album';

            $e                    = $this->GetTitleDesc($filename, false);
            $oneItem->title       = $e->title;
            $oneItem->description = $e->description;

            $oneItem->albumID = rawurlencode($this->albumID);
            if ($this->albumID == '0' || $this->albumID == '') {
                // $oneItem->ID = $this->CustomEncode($filename);
                $oneItem->ID = rawurlencode($this->CustomEncode($filename));
            } else {
                // $oneItem->ID = $this->albumID . $this->CustomEncode('/' . $filename);
                $oneItem->ID = rawurlencode($this->albumID . $this->CustomEncode('/' . $filename));
            }

            $s = $this->GetAlbumCover($this->data->fullDir . $filename . '/');
            if ($s != '') {
                // a cover has been found
                $path = '';
                if ($this->albumID == '0') {
                    $path = $filename;
                } else {
                    $path = $this->album . '/' . $filename;
                }
                $oneItem->srct       = rawurlencode($this->CustomEncode($this->config['contentFolder'] .'/'. $path . '/' . $s));
                // $oneItem->srct       = $this->CustomEncode(self::CONTENT_FOLDER .'/'. $path . '/' . $s);
                $size                = getimagesize(__DIR__ . $this->config['contentFolder'] . '/' . $path . '/' . $s);
                $oneItem->imgtWidth  = $size[0];
                $oneItem->imgtHeight = $size[1];
                return $oneItem;
            }
        }
    }

}
?>